[BB-3306] Customizable Certificate Date Format

shimulch opened the pull request:
The current way of defining Certificate Date Format doesn't allow easy override. That sometimes leads to code drift when trying to [customize](https://github.com/edx/edx-platform/commit/f360a39cfa03f53458540a0acd4fbe997865a97c). This PR introduces a new setting ``CERTIFICATE_DATE_FORMAT``, which allows easy customization of date format in certificate.

**JIRA tickets**: https://tasks.opencraft.com/browse/BB-3306

~~**Discussions**:~~

**Dependencies**: None

~~**Screenshots**:~~

~~**Sandbox URL**:~~

**Merge deadline**: None

**Testing instructions**:

1. In your local devstack check certificate date format.
2. Clone this PR on local devstack.
3. To enable certificate, add a new [HTML View Configuration](https://edx.readthedocs.io/projects/edx-installing-configuring-and-running/en/latest/configuration/enable_certificates.html#configure-course-certificates-for-your-open-edx-instance) from ``/admin/certificates/certificatehtmlviewconfiguration/``. Make sure to mark it as enabled.
4. Check if your course has appropriate ``Course Mode`` here - ``/admin/course_modes/coursemode/``. If you are testing with ``course-v1:edX+DemoX+Demo_Course``, it should be fine.
5. Go to Studio and set up the certificate. For example, go to ``http://localhost:18010/certificates/course-v1:edX+DemoX+Demo_Course``, and click the ``Setup Your Certificate`` button.
6. Fill up the form with dummy info and click ``Create``.
7. You should see a ``Preview Certificate`` button.
8. The date of our interest can be seen at the bottom right corner of the certificate - ``Issued On:``
3. Check that certificates Issued date format.
4. Override ``CERTIFICATE_DATE_FORMAT`` with new date format. The format string should be supported by [strftime_localized](https://github.com/edx/edx-platform/blob/master/common/djangoapps/util/date_utils.py#L105)
5. Check the certificate to observe the change.

**Author notes and concerns**:
N/A

**Reviewers**
- [x] @farhaanbukhsh 
- [ ] edX reviewer[s] TBD

edx-webhook commented:
Thanks for the pull request, @shimulch! I've created [OSPR-5288](https://openedx.atlassian.net/browse/OSPR-5288) to keep track of it in JIRA, where we prioritize reviews. Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


This is currently a draft pull request. When it is ready for our review and all tests are green, click "Ready for Review", or remove "WIP" from the title, as appropriate.

<!-- comment:end_of_wip -->




<!-- comment:external_pr -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IHRydWV9 -->

farhaanbukhsh commented on lms/envs/common.py:
```diff
@@ -1047,6 +1047,8 @@ def _make_mako_template_dirs(settings):
     'url_root': '/static/django-pyfs',
 }
 
+CERTIFICATE_DATE_FORMAT = "%B %-d, %Y"
```
It will be good to add comments explaining the format supported here and also a comment explaining the constant.

shimulch replied:
Added docs related to use of this new configuration.

farhaanbukhsh reviewed (changes requested):
I think it will be good to add comments on the constant, what do you think?

Also just a nit but if you are adding commit message start with a capital letter and keep the message imperative. I find this blog really good https://chris.beams.io/posts/git-commit/

shimulch commented:
@farhaanbukhsh, Thanks for your review. Yes, I've added docs on what formats it would support. Also, I've updated the test instructions.

And, thanks for the tip regarding the commit message, I will keep that in mind from now on. That's a nice thoughtfull article. :-)

farhaanbukhsh commented on lms/envs/common.py:
```diff
@@ -1047,6 +1047,11 @@ def _make_mako_template_dirs(settings):
     'url_root': '/static/django-pyfs',
 }
 
+# Certificate issued date format. Supports all formats supported by
+# ``common.djangoapps.util.date_utils.strftime_localized``, which
+# is an extension over python ``strftime`` function.
```
```suggestion
#  Set certificate issued date format. It supports all formats supported by
# ``common.djangoapps.util.date_utils.strftime_localized``.
```

farhaanbukhsh reviewed (approved):
👍

We can just change the wording and squash the commits to one and it's good to go from myside.

- [x] I tested this: Generated the certificate and changed and tested different date formats.
- [x] I read through the code
~~- [ ] I checked for accessibility issues~~
- [x] Includes documentation

natabene commented:
@shimulch Thank you for your contribution. Please let me know once it is ready for our review.

shimulch commented:
@natabene, It's ready for edX review. :-)

natabene commented:
@shimulch Thank you, we intend to review it within the next 2 weeks.

tuchfarber commented:
This looks pretty straightforward to me. I'm going to merge and I'll post again once I've confirmed it hasn't caused any weird production-level issues.

edx-webhook commented:
@shimulch 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/25777&entry.1484655318=https://github.com/shimulch&entry.752974735=2020-12-07+01:53&entry.1917517419=2021-01-27+21:01&entry.2133058324=Yes) so we can improve your experience in the future.

tuchfarber commented:
Clearly this went out alright ^ Haha. Just wanted to says thanks for the easy PR and excellent testing instructions.
